pkg/registry/core/service/rest.go

```diff
@@ -432,19 +435,55 @@ func (rs *REST) ResourceLocation(ctx genericapirequest.Context, id string) (*url
 		}
 		for i := range ss.Ports {
 			if ss.Ports[i].Name == portStr {
-				// Pick a random address.
-				ip := ss.Addresses[rand.Intn(len(ss.Addresses))].IP
-				port := int(ss.Ports[i].Port)
-				return &url.URL{
-					Scheme: svcScheme,
-					Host:   net.JoinHostPort(ip, strconv.Itoa(port)),
-				}, rs.proxyTransport, nil
+                                // This is a little wonky, but it's expensive to test for the presence of a Pod
+                                // So we repeatedly try at random and validate it, this means that for an invalid
+                                // service with a lot of endpoints we're going to potentially make a lot of calls,
+                                // but in the expected case we'll only make one. 
+				for try := 0; try < len(ss.Addresses); try++ {
+					// Pick a random address.
+					addr := ss.Addresses[rand.Intn(len(ss.Addresses))]
+					if !utilfeature.DefaultFeatureGate.Enabled(features.ServiceProxyAllowExternalIPs) {
+						if err := isValidAddress(ctx, &addr, rs.pods); err != nil {
+							glog.Errorf("Address %v isn't valid (%v)", addr, err)
+							continue
+						}
+					}
+					ip := addr.IP
+					port := int(ss.Ports[i].Port)
+					return &url.URL{
+						Scheme: svcScheme,
+						Host:   net.JoinHostPort(ip, strconv.Itoa(port)),
+					}, rs.proxyTransport, nil
+				}
+				glog.Errorf("Failed to find a valid address, skipping subset: %v", ss)
 			}
 		}
 	}
 	return nil, nil, errors.NewServiceUnavailable(fmt.Sprintf("no endpoints available for service %q", id))
 }
 
+func isValidAddress(ctx genericapirequest.Context, addr *api.EndpointAddress, pods *podstore.REST) error {
+	if addr.TargetRef == nil {
+		return fmt.Errorf("Address has no target ref, skipping: %v", addr)
+	}
+	nsCtx := genericapirequest.WithNamespace(ctx, addr.TargetRef.Namespace)
```

Comment:
we already have a context passed in with the right namespace, can use that (and should verify the addr.TargetRef.Namespace matches the namespace in the context... no cross-namespace pod references in an endpoint)